File: src/main/java/TestRecursion.java
/**
 * @author 赵世敏
 * @date 2022/7/2
 * 1208676641@qq.com
 */
public class TestRecursion {
    public static void main(String[] args) {
        //计算阶乘
        System.out.println(factorial(6));
        System.out.println(factorial1(6));
    }
    //循环实现
    public static int factorial(int n){
        int result = 1;
        for (int i=1 ;i<=n ;i++){
            result *= i;
        }
        return result;
    }
    //递归实现
    public static int factorial1(int n){
        //递归结束条件
        if (n==1){
            return 1;
        }
        return n*factorial1(n-1);
    }
}

File: src/main/scala/function/TestRecursion.scala
package function

/**
 * @author 赵世敏
 * @date 2022/7/2
 *       1208676641@qq.com
 *
 */
object TestRecursion {
  def main(args: Array[String]): Unit = {
    println(factorial(5))
    println(tailFact(5))
  }

  //递归实现阶乘
  def factorial(n: Int): Int = {
    if (n == 1) return 1
    factorial(n - 1) * n
  }
  //尾递归实现
  def tailFact(n:Int):Int={
    @annotation.tailrec
    def loop(n:Int,recursionResult:Int):Int={
      if(n==1) recursionResult
      else loop(n-1,n*recursionResult)
    }
    loop(n,1)
  }

}
